### PR TITLE
Fix issue 65 operator hiding

### LINKED ISSUE #65: a C++ derived class that defines an operator hides automatic base class conversion
@Syniurge : a C++ derived class that defines an operator hides automatic base class conversion

test.h:
```
struct A{ int x;};

struct B : public A{
  // comment to remove error
  operator float() const;
};
```

main.d:
```
modmap (C++) "test.h";
import (C++) A;
import (C++) B;

void main(){
  B b;
  int bx=b.x;
}
```

compiling produces this error:
`Error: template instance opCast!(A) does not match template declaration opCast(type : float)`

The error goes away if we remove ` operator float() const;`

It works if we add `operator A*() {return this;}` in `B` and change the D code to:
`int bx=(cast(A*)b).x;` but this is obviously not practical.

### PR DESCRIPTION
**Pull request by timotheecour**
fixes https://github.com/Syniurge/Calypso/issues/65
I wonder if there's a better way, and if it could have negative side effects

NOTE: depends on https://github.com/Syniurge/Calypso/pull/93 , which should be merged first

### CODE REVIEW

**Comment by Syniurge** [comment]
Sorry for not commenting earlier on #65, I'll look into it tonight.

But most likely we shouldn't be gagging for all overloading casts into opCast calls between C++ struct/classes, there may be conversion `operator MyClass()` that we want to map and use in D (e.g for OpenCV iirc there are multiple `cv::Point_<>` classes that can be casted between each other).

Your PR seems to be on the right track but this shouldn't be specific to C++ aggregates, but to all struct/class values, i.e checked with `isAggregateValue()` (and excepted if `to.isBaseOf(from)` is true).

**Comment by timotheecour** [comment]
>  there may be conversion operator MyClass() that we want to map and use in D (e.g for OpenCV iirc there are multiple cv::Point_<> classes that can be casted between each other).

I added a test case that shows that `{auto a2=cast(double)A.init;}` compiles, ie that `operator double() const {return 1.5;}` is indeed mapped.

But maybe I'm misunderstanding what you're saying; do you have a test case in mind that breaks?

> Your PR seems to be on the right track but this shouldn't be specific to C++ aggregates, but to all struct/class values, i.e checked with isAggregateValue() (and excepted if to.isBaseOf(from) is true).

I'm trying to only add special handling when the type is C++ struct mapped to D via calypso, other cases (eg a class, which is an isAggregateValue according to that function definition) should be using the regular D rules. Again, do you have a failing test case in mind?

* what's a good, menomonic name for that thing? (C++ struct mapped to D via calypso); to distinguish from extern(C++) struct/class which is another thing

**Comment by Syniurge** [comment]
> But maybe I'm misunderstanding what you're saying;

Disregard my previous message, I barely understand myself either. What I wanted to say was:

1. Error gagging should generally be avoided
2. There may be erroneous casts between C++ aggregates where we want the `template instance opCast!(...) does not match template declaration opCast(...)` error message to appear
3. The exception to operator overloading should be if `to` is a base aggregate of `from`
4. This should not be specific to C++ aggregates, the logic is the same for any language featuring multiple inheritance.

After having a look at the code, the `CastExp` is actually created by Calypso's `getRightThis()` in **cppaggregate.cpp**, and this `CastExp` prompts codegen to adjust pointers if needed.

There is also another plain bug in `CastExp.semantic` right after:
```d
        if (tob.ty == Tstruct && !tob.equals(t1b))
        {
            /* Look to replace:
             *  cast(S)t
             * with:
             *  S(t)
             */

            // Rewrite as to.call(e1)
            Expression e = new TypeExp(loc, to);
            e = new CallExp(loc, e, e1);
            e = e.trySemantic(sc);
            if (e)
                return e;
        }
```
With Calypso `StructDeclaration` might now be bases of `ClassDeclaration`, so we probably want the same exception to kick in here.

(I don't seem to have tested struct as bases much, I'm seeing other obvious bugs in dcast.d)

File: runtime/CMakeLists.txt [hunk 1]
```diff
@@ -478,7 +478,8 @@ macro(build_runtime_libs druntime_o druntime_bc phobos2_o phobos2_bc calypsort_o
         LIBRARY_OUTPUT_DIRECTORY    ${output_path}
         RUNTIME_OUTPUT_DIRECTORY    ${output_path}
         COMPILE_FLAGS               "${c_flags}"
-        LINK_FLAGS                  "${ld_flags}"
+        # on OSX at least, druntime needed to avoid link error
+        LINK_FLAGS                  "${ld_flags} -L../lib -ldruntime-ldc${lib_suffix} "
```
Comment (Syniurge) [inline runtime/CMakeLists.txt:482]: hmm is this related to the PR/issue?

Comment (timotheecour) [reply runtime/CMakeLists.txt:482]: see note on top: `NOTE: depends on #93 , which should be merged first`

**Comment by Syniurge** [comment]
@timotheecour I pushed a fix that simply disables overloading by opCast when the cast is from a derived class to a base aggregate or vice versa. Thanks for taking the time to investigate and for the original fix.
